added orthogonal initializer

mikowals opened the pull request:
Added an orthogonal initialiser like the one in [TensorFlow python](https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/ops/init_ops.py#L548).

Is there a swift test similar to `assert_allclose` to use for equality with tolerance for small differences?

saeta commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,24 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+        let columnCount = shape[shape.rank - 1]
+        let flatShape: TensorShape = rowCount < columnCount ? [columnCount, rowCount] : [rowCount, columnCount]
+        let normal = Tensor(randomNormal: flatShape, seed: seed)
+        var (q, r) = Raw.qr(normal, fullMatrices: false)
```
Does it make sense to add a `qr` method to Tensor?

mikowals replied:
I added both this and the diagPart method mentioned below.

saeta commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,24 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+        let flatShape: TensorShape = rowCount < columnCount ? [columnCount, rowCount] : [rowCount, columnCount]
+        let normal = Tensor(randomNormal: flatShape, seed: seed)
+        var (q, r) = Raw.qr(normal, fullMatrices: false)
+        let d = Raw.diagPart(r)
```
Same for `diagPart`.

dan-zheng commented:
> Is there a swift test similar to `assert_allclose` to use for equality with tolerance for small differences?

Yes, check out the `assertEqual` functions defined in [`Tests/TensorFlowTests/Helpers.swift`](https://github.com/tensorflow/swift-apis/blob/master/Tests/TensorFlowTests/Helpers.swift).

saeta commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,24 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
 }
+
+public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    init(
```
Please add some documentation comments here to describe this initializer.

saeta commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,24 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    init(
+        orthogonal shape: TensorShape,
+        gain: Scalar = 1,
+        seed: (Int32, Int32) = (Int32.random(in: Int32.min..<Int32.max),
```
For usability, we've been making the Tensor API take `Int`'s wherever possible. (Note: this will require you to do the appropriate casting within the initializer in Swift between the variable- and fixed-width number formats.)

rxwei replied:
This is actually not what we are doing for random seeds. We only use `Int` for shape dimensions and count-like numbers. Random seeds are backend specific (in this case, TensorFlow graph specific). Other initializers are also using `Int32` seeds. Usability problems are small for random seeds because they are usually literals and are not part of complicated data flow.

mikowals replied:
based on @rxwei comment I left this unchanged.

saeta commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,24 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+        } 
+        self = q.reshaped(to: shape) * gain 
+    }
+}
```
Formatting nit: newline.

mikowals replied:
added.

saeta commented on Tests/TensorFlowTests/InitializerTests.swift:
```diff
@@ -71,13 +71,32 @@ final class InitializerTests: XCTestCase {
         ("testArrayConversion", testArrayConversion),
         ("testDataTypeCast", testDataTypeCast),
-        ("testBoolToNumericCast", testBoolToNumericCast)
+        ("testBoolToNumericCast", testBoolToNumericCast),
```
Thanks for catching and adding this one! :-)

mikowals commented:
I have added the methods qr() and diagPart() in the linear algebra section of math.swift.  I can add a test for each of these also.  I am agnostic about whether having these methods is better than just calling their `Raw.` versions.

I added documentation and I think I have addressed all the comments. I also changed the test to use `assertEqual`.

mikowals commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,24 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+        let normal = Tensor(randomNormal: flatShape, seed: seed)
+        var (q, r) = Raw.qr(normal, fullMatrices: false)
+        let d = Raw.diagPart(r)
+        q *= Raw.sign(d)
```
You are correct and I am now using the top-level `sign()`

eaplatanios commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2158,6 +2158,36 @@ internal func _vjpMatmul<Scalar: TensorFlowFloatingPoint>(
     })
 }
 
+public extension Tensor where Scalar: FloatingPoint & TensorFlowScalar {
```
```suggestion
public extension Tensor where Scalar: TensorFlowFloatingPoint {
```

mikowals replied:
done.

eaplatanios commented on Tests/TensorFlowTests/InitializerTests.swift:
```diff
@@ -71,13 +71,32 @@ final class InitializerTests: XCTestCase {
+            XCTAssertEqual(shape, t.shape.dimensions)
+        
+            // Check orthogonality by computing the inner product
+            t = t.reshaped(to: [t.shape.dimensions[0 ..< t.rank-1].reduce(1, *), t.shape[t.rank - 1]])
```
```suggestion
            t = t.reshaped(to: [t.shape.dimensions[0..<(t.rank - 1)].reduce(1, *), t.shape[t.rank - 1]])
```

mikowals replied:
made suggested changes.

eaplatanios commented on Tests/TensorFlowTests/OperatorTests/MathTests.swift:
```diff
@@ -394,6 +436,7 @@ final class MathOperatorTests: XCTestCase {
         ("testXWPlusB", testXWPlusB),
         ("testXORInference", testXORInference),
         ("testMLPClassifierStruct", testMLPClassifierStruct),
+        ("testQRApproximation", testQRApproximation),
```
`testDiagPart` is missing here.

mikowals replied:
added it.

eaplatanios commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,39 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    ///
+    /// - Parameters:
+    ///   - shape: The dimensions of the tensor.
+    ///   - gain: A multiplicative factor to apply to the orthogonal tensor.
```
Could you also document `seed`?

mikowals replied:
added seed to documentation

eaplatanios commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,39 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+        seed: (Int32, Int32) = (Int32.random(in: Int32.min..<Int32.max),
+                                Int32.random(in: Int32.min..<Int32.max))
+    ) {
+        let rowCount = shape[0 ..< shape.rank - 1].dimensions.reduce(1, *)
```
```suggestion
        let rowCount = shape[0..<(shape.rank - 1)].dimensions.reduce(1, *)
```

mikowals replied:
made suggested change.

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2158,6 +2158,36 @@ internal func _vjpMatmul<Scalar: TensorFlowFloatingPoint>(
+    ///   - fullMatrices: If true, compute full-sized `q` and `r`. If false
+    ///     (the default), compute only the leading `P` columns of `q`.
+    ///  
+    func qr(fullMatrices: Bool = false) -> (Tensor<Scalar>, Tensor<Scalar>) {
```
Avoid abbreviations and use terms of art. See [Swift API Design Guidelines - Use Terminology Well](https://swift.org/documentation/api-design-guidelines/#use-terminology-well) for more details.
```suggestion
    func qrDecomposition(fullMatrices: Bool = false) -> (q: Tensor<Scalar>, r: Tensor<Scalar>) {
```

mikowals replied:
Done.

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
````diff
@@ -2158,6 +2158,36 @@ internal func _vjpMatmul<Scalar: TensorFlowFloatingPoint>(
+    /// t.diagPart() ==> [1, 2, 3, 4]
+    /// ```
+    ///
+    func diagPart() -> Tensor<Scalar> {
````
```suggestion
    func diagonalPart() -> Tensor<Scalar> {
```

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
````diff
@@ -2158,6 +2158,36 @@ internal func _vjpMatmul<Scalar: TensorFlowFloatingPoint>(
+    /// For example:
+    ///
+    /// ```
+    /// # 't' is [[1, 0, 0, 0]
````
Use Swift code in code snippets. Swift comments start with `//`.
```suggestion
    /// // 't' is [[1, 0, 0, 0]
```

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2158,6 +2158,36 @@ internal func _vjpMatmul<Scalar: TensorFlowFloatingPoint>(
+    ///           [0, 2, 0, 0]
+    ///           [0, 0, 3, 0]
+    ///           [0, 0, 0, 4]]
+    ///
```
I think this empty line can be removed.
```suggestion
```

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2158,6 +2158,36 @@ internal func _vjpMatmul<Scalar: TensorFlowFloatingPoint>(
 }
 
+public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// Computes the QR decomposition of each inner matrix in `tensor` such that
```
```suggestion
    /// Returns the QR decomposition of each inner matrix in the tensor such that
```
- #160
- `tensor` is an undefined identifier. Use "the tensor" instead.

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2158,6 +2158,36 @@ internal func _vjpMatmul<Scalar: TensorFlowFloatingPoint>(
 
+public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// Computes the QR decomposition of each inner matrix in `tensor` such that
+    /// `tensor[..., :, :] = q[..., :, :] * r[..., :,:])`
```
The syntax appears to be Python, and thus could be confusing to the user. We should use Swift code in all code examples. However, our APIs don't fully support this concise indexing style. For now, I believe it is more clear to use an English sentence to describe QR decomposition. This way you can also talk about the matrix properties of the results like orthogonality.

> Returns the QR decomposition of each inner matrix in the tensor, a tensor with inner orthogonal matrices `q` and a tensor with inner upper triangular matrices `r`, such that the tensor is equal to `matmul(q, r)`.

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
````diff
@@ -2158,6 +2158,36 @@ internal func _vjpMatmul<Scalar: TensorFlowFloatingPoint>(
+    ///
+    /// ```
+    /// # 't' is [[1, 0, 0, 0]
+    ///           [0, 2, 0, 0]
````
```suggestion
    /// //         [0, 2, 0, 0]
```

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
````diff
@@ -2158,6 +2158,36 @@ internal func _vjpMatmul<Scalar: TensorFlowFloatingPoint>(
+    /// ```
+    /// # 't' is [[1, 0, 0, 0]
+    ///           [0, 2, 0, 0]
+    ///           [0, 0, 3, 0]
````
```suggestion
    /// //         [0, 0, 3, 0]
```

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2158,6 +2158,36 @@ internal func _vjpMatmul<Scalar: TensorFlowFloatingPoint>(
+    /// # 't' is [[1, 0, 0, 0]
+    ///           [0, 2, 0, 0]
+    ///           [0, 0, 3, 0]
+    ///           [0, 0, 0, 4]]
```
```suggestion
    /// //         [0, 0, 0, 4]]
```

rxwei commented on Tests/TensorFlowTests/OperatorTests/MathTests.swift:
```diff
@@ -356,6 +356,48 @@ final class MathOperatorTests: XCTestCase {
+        let target1 = Tensor<Float>([1, 6, 11, 16])
+        XCTAssertEqual(target1, t1.diagPart())
+
+        //test on 4D tensor
```
Some minor nitpick:
- Add a space between `//` and a comment. 
- Start each comment with an upper case letter.
- End each comment with a period.
- We've been consistently adding a dash between a number and "D" when they denote dimensionality in comments.
```suggestion
        // Test on a 4-D tensor.
```

rxwei commented on Tests/TensorFlowTests/OperatorTests/MathTests.swift:
```diff
@@ -356,6 +356,48 @@ final class MathOperatorTests: XCTestCase {
+            let aReconstitutedFull = matmul(qFull, rFull)
+            assertEqual(a, aReconstitutedFull, accuracy: 1e-5)
+        }
+        
```
Remove empty line.

rxwei commented on Tests/TensorFlowTests/InitializerTests.swift:
```diff
@@ -71,13 +71,32 @@ final class InitializerTests: XCTestCase {
+            XCTAssertEqual(shape, t.shape.dimensions)
+        
+            // Check orthogonality by computing the inner product
+            t = t.reshaped(to: [t.shape.dimensions[0..<(t.rank - 1)].reduce(1, *), t.shape[t.rank - 1]])
```
Fit within 100 columns.

rxwei commented on Tests/TensorFlowTests/InitializerTests.swift:
```diff
@@ -71,13 +71,32 @@ final class InitializerTests: XCTestCase {
     }
 
+    func testOrthogonalShapesValues() {
+        for shape in [[10, 10], [10, 9, 8], [100, 5, 5], [50, 40], [3, 3, 32, 64]]{
```
Add a space.
```suggestion
        for shape in [[10, 10], [10, 9, 8], [100, 5, 5], [50, 40], [3, 3, 32, 64]] {
```

rxwei commented on Tests/TensorFlowTests/InitializerTests.swift:
```diff
@@ -71,13 +71,32 @@ final class InitializerTests: XCTestCase {
 
+    func testOrthogonalShapesValues() {
+        for shape in [[10, 10], [10, 9, 8], [100, 5, 5], [50, 40], [3, 3, 32, 64]]{
+            // Check the shape
```
End each comment with a period.
```suggestion
            // Check the shape.
```

rxwei commented on Tests/TensorFlowTests/InitializerTests.swift:
```diff
@@ -71,13 +71,32 @@ final class InitializerTests: XCTestCase {
+            var t = Tensor<Float>(orthogonal: TensorShape(shape))
+            XCTAssertEqual(shape, t.shape.dimensions)
+        
+            // Check orthogonality by computing the inner product
```
```suggestion
            // Check orthogonality by computing the inner product.
```

rxwei commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,40 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    ) {
+        let rowCount = shape[0..<(shape.rank - 1)].dimensions.reduce(1, *)
+        let columnCount = shape[shape.rank - 1]
+        let flatShape: TensorShape = rowCount < columnCount ? [columnCount, rowCount] : [rowCount, columnCount]
```
Fit within 100 columns.

rxwei commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,40 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
 }
+
+public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// Initializer that generates an orthogonal matrix or tensor. 
```
Initializer documentation should start with `Creates a ...`. In this case, it should be `Creates an orthogonal matrix or tensor.`

rxwei commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,40 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// have orthogonal rows. Otherwise, the output will have orthogonal columns.
+    /// 
+    /// If the shape of the tensor to initialize is more than two-dimensional, a matrix of shape 
+    /// (shape[0] * ... * shape[n - 2], shape[n - 1]) is initialized, where n is the length of the 
```
- Use backquotes around any code in comments.
- Use square brackets `[` and `]` for shapes to match the array literal syntax.
- `n` is undefined. Did you mean the `rank` property?
```suggestion
    /// `[shape[0] * ... * shape[rank - 2], shape[rank - 1]]` is initialized, where n is the length of the 
```

rxwei commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,40 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// shape vector. The matrix is subsequently reshaped to give a tensor of the desired shape.
+    ///
+    /// - Parameters:
+    ///   - shape: The dimensions of the tensor.
```
```suggestion
    ///   - shape: The shape of the tensor.
```

rxwei commented on Sources/TensorFlow/Initializers.swift:
```diff
@@ -548,3 +548,40 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+        seed: (Int32, Int32) = (Int32.random(in: Int32.min..<Int32.max),
+                                Int32.random(in: Int32.min..<Int32.max))
+    ) {
+        let rowCount = shape[0..<(shape.rank - 1)].dimensions.reduce(1, *)
```
```suggestion
        let rowCount = shape.dimensions.dropLast().reduce(1, *)
```

mikowals commented:
I have made the requested changes.

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2181,3 +2181,34 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
+public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// Returns the QR decomposition of each inner matrix in the tensor, a tensor with inner 
+    /// orthogonal matrices q and a tensor with inner upper triangular matrices r, such that the 
+    /// tensor is equal to matmul(q, r).
```
Use backquotes around any code.
```suggestion
    /// tensor is equal to `matmul(q, r)`.
```

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2181,3 +2181,34 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
+
+public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// Returns the QR decomposition of each inner matrix in the tensor, a tensor with inner 
+    /// orthogonal matrices q and a tensor with inner upper triangular matrices r, such that the 
```
```suggestion
    /// orthogonal matrices `q` and a tensor with inner upper triangular matrices `r`, such that the 
```

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2181,3 +2181,34 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// //         [0, 2, 0, 0]
+    /// //         [0, 0, 3, 0]
+    /// //         [0, 0, 0, 4]]
+    /// // t.diagPart() ==> [1, 2, 3, 4]
```
```suggestion
    /// t.diagonalPart()
    /// // [1, 2, 3, 4]
```

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2181,3 +2181,34 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// 
+    /// - Parameters:
+    ///   - fullMatrices: If true, compute full-sized `q` and `r`. If false
+    ///     (the default), compute only the leading `P` columns of `q`.
```
- Identifier `P` is undefined.
- Could you help make sure this documentation item fits to size (100 columns)?
```suggestion
    ///     (the default), compute only the leading `min(shape[rank - 1], shape[rank - 2])` columns of `q`.
```

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -2181,3 +2181,34 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// tensor is equal to matmul(q, r).
+    /// 
+    /// - Parameters:
+    ///   - fullMatrices: If true, compute full-sized `q` and `r`. If false
```
```suggestion
    ///   - fullMatrices: If `true`, compute full-sized `q` and `r`. Otherwise,
```

mikowals commented:
Updated with the latest documentation fixes.
